components/nimbus/src/enrollment.rs

```diff
@@ -597,86 +599,190 @@ impl<'a> EnrollmentsEvolver<'a> {
+            // New experiment.
+            (None, Some(experiment), None) => Some(ExperimentEnrollment::from_new_experiment(
+                is_user_participating,
+                self.nimbus_id,
+                self.available_randomization_units,
+                self.app_context,
+                experiment,
+                out_enrollment_events,
+            )?),
+            // Experiment deleted remotely.
+            (Some(_), None, Some(enrollment)) => {
+                enrollment.on_experiment_ended(out_enrollment_events)
+            }
+            // Known experiment.
+            (Some(_), Some(experiment), Some(enrollment)) => {
+                Some(enrollment.on_experiment_updated(
                     is_user_participating,
                     self.nimbus_id,
                     self.available_randomization_units,
                     self.app_context,
                     experiment,
                     out_enrollment_events,
-                )?),
-                // Experiment deleted remotely.
-                (Some(_), None, Some(enrollment)) => {
-                    enrollment.on_experiment_ended(out_enrollment_events)
-                }
-                // Known experiment.
-                (Some(_), Some(experiment), Some(enrollment)) => {
-                    Some(enrollment.on_experiment_updated(
-                        is_user_participating,
-                        self.nimbus_id,
-                        self.available_randomization_units,
-                        self.app_context,
-                        experiment,
-                        out_enrollment_events,
-                    )?)
-                }
-                (None, None, Some(enrollment)) => enrollment.maybe_garbage_collect(),
-                (None, Some(_), Some(_)) => {
-                    return Err(NimbusError::InternalError(
-                        "New experiment but enrollment already exists.",
-                    ))
-                }
-                (Some(_), None, None) | (Some(_), Some(_), None) => {
-                    return Err(NimbusError::InternalError(
-                        "Experiment in the db did not have an associated enrollment record.",
-                    ))
-                }
-                (None, None, None) => unreachable!(),
-            },
-        )
+                )?)
+            }
+            (None, None, Some(enrollment)) => enrollment.maybe_garbage_collect(),
+            (None, Some(_), Some(_)) => {
+                return Err(NimbusError::InternalError(
+                    "New experiment but enrollment already exists.",
+                ))
+            }
+            (Some(_), None, None) | (Some(_), Some(_), None) => {
+                return Err(NimbusError::InternalError(
+                    "Experiment in the db did not have an associated enrollment record.",
+                ))
+            }
+            (None, None, None) => unreachable!(),
```

Comment:
The code has been this way for a while, I think, but this is a good catch: the code isn't tehcnically unreachable in the sense that if we get there, evolve_enrollment has been called with parameters that don't make any sense together: there's no prev_experiment, no next_experiment, and no prev_enrollment, meaning that there's nothing that could evolve or be evolved.

Presumably this could happen due to bugs in the system somewhere (most likely a bug or unhandled case somewhere else in the SDK).

That said, unreachable! causes the client program to panic, which seems like an unreasonable thing to do on behalf of an embedding client.  I'd be more inclined to just return an Err.  @travis79 @jhugman thoughts?

Reply:
In the interest of not crashing the embedding app, I think an error seems appropriate.